chat-plugins/poll.js

```diff
@@ -0,0 +1,158 @@
+			this.options.set(i + 1, {name: options[i], votes: 0});
+		}
+	}
+
+	Poll.prototype.vote = function (user, option) {
+		if (this.voters.has(user.latestIp)) {
+			return user.sendTo(this.room, "You have already voted for this poll.");
+		} else {
+			this.voters.add(user.latestIp);
+		}
+
+		this.options.get(option).votes++;
+		this.totalVotes++;
+
+		this.update();
+	};
+
+	Poll.prototype.generateVotes = function () {
+		var output = '<div class="infobox" style="text-align:center"><p style="font-weight:bold;font-size:14pt">' + this.question + '</p>';
+		this.options.forEach(function (option, number) {
+			output += '<button value="/votepoll ' + number + '" name="send">' + number + '. ' + Tools.escapeHTML(option.name) + '</button><br/>';
+		});
+		output += '</div>';
+
+		return output;
+	};
+
+	Poll.prototype.generateResults = function () {
+		var output = '<div class="infobox" style="text-align:center"><p style="font-weight:bold;font-size:14pt">' + this.question + '</p>';
+		var iter = this.options.entries();
+
+		var i = iter.next();
+		while (!i.done) {
+			output += '<span style="font-weight:bold;padding:2px">' + i.value[0] + '. ' + Tools.escapeHTML(i.value[1].name) + '</span><br/><span style="font-size:8pt;padding:0px">' + i.value[1].votes + ' votes (' + Math.round((i.value[1].votes * 100) / this.totalVotes) + '% of total).</span><br/>';
+			i = iter.next();
+		}
+		output += '</div>';
+
+		return output;
+	};
+
+	Poll.prototype.update = function () {
+		var results = this.generateResults();
+
+		// Update the poll results for everyone that has voted
+		for (var i in this.room.users) {
+			var user = this.room.users[i];
+			if (this.voters.has(user.latestIp)) {
+				user.sendTo(this.room, '|uhtmlchange|poll' + this.room.pollNumber +  '|' + results);
+			}
+		}
+	};
+
+	Poll.prototype.display = function (user, broadcast) {
+		var votes = this.generateVotes();
+		var results = this.generateResults();
+
+		var target = {};
+
+		if (broadcast) {
+			target = this.room.users;
+		} else {
+			target[0] = user;
+		}
+
+		for (var i in target) {
+			var thisUser = target[i];
+			if (this.voters.has(thisUser.latestIp)) {
+				thisUser.sendTo(this.room, '|uhtml|poll' + this.room.pollNumber +  '|' + results);
+			} else {
+				thisUser.sendTo(this.room, '|uhtml|poll' + this.room.pollNumber +  '|' + votes);
+			}
+		}
+	};
+
+	Poll.prototype.end = function () {
+		var results = this.generateResults();
+
+		this.room.send('|uhtml|poll' + this.room.pollNumber +  '|' + results);
+		this.room.send('|uhtmlchange|poll' + this.room.pollNumber +  '|' + results);
+	};
+
+	return Poll;
+})();
+
+exports.commands = {
+	newpoll: function (target, room, user) {
+		if (!this.can('ban', null, room)) return false;
+		if (room.poll) return this.sendReply("There is already a poll in progress in this room.");
+
+		var targets = target.split(target.includes('|') ? '|' : ',');
+		if (targets.length < 3) {
+			return this.sendReply("Not enough arguments for /newpoll.");
```

Comment:
Bumbadadabum, if you're going to update the code, I recommend you to change error parts into: ``this.errorReply`` instead of ``this.sendReply``.